### PR TITLE
Allow providers to defer trainees who haven't started

### PR DESCRIPTION
**Pull request by ralph-hawkins**
Trainees can now be deferred if their course start date is in the future.

localhost_3000_record_2a1eff12-ce41-4c3e-a42b-4aeae604fbdd(screenshot) (1)

Adds start date to deferral confirmation (regardless of when the trainee started). This feels relevant here and might help prevent some issues around trainees who had a predicted start date being withdrawn.

localhost_3000_record_2a1eff12-ce41-4c3e-a42b-4aeae604fbdd_defer_confirm_referrer=_record_2a1eff12-ce41-4c3e-a42b-4aeae604fbdd(screenshot) (1)

localhost_3000_record_2a1eff12-ce41-4c3e-a42b-4aeae604fbdd_defer_confirm_referrer=_record_2a1eff12-ce41-4c3e-a42b-4aeae604fbdd(screenshot) (2)

If the trainee didn't start their ITT before deferring the record shows that "The trainee deferred before starting their ITT" and the start date cannot be edited after a trainee has been withdrawn.

localhost_3000_record_2a1eff12-ce41-4c3e-a42b-4aeae604fbdd(screenshot) (1)

If a trainee is reinstated who didn't previously have a start date, the reinstatement date becomes their start date too.

### Bonus

Fixed the line length in the outstanding actions banner:

localhost_3000_record_2a1eff12-ce41-4c3e-a42b-4aeae604fbdd(screenshot) (4)

### CODE REVIEW

File: app/lib/utils.js [hunk 1]
```diff
@@ -967,31 +967,42 @@ exports.needsCourseDates = record => {
...
-  - ITT start date is in the future
-  - the ITT has started but there is no trainee start date
-    (but we need to check if trainee started - see
-    needsTraineeStarDate)
+  true if itt start date is in the future
   ====================================================================
+*/
 
-  Usage:
+exports.ittInTheFuture = (record) => {
```
Comment (edwardhorsford) [inline app/lib/utils.js:985]: This filter seems to be necessary? It's just going to return the same as the existing `isInFuture` filter isn't it?

File: app/lib/utils.js [hunk 2]
```diff
@@ -1000,42 +1011,43 @@ exports.showDelete = (record) => {
...
+  true if course start is in the past AND
+  trainee has a commencement date
   ====================================================================
 */
 
-exports.needTraineeStartDate = (record) => {
-  let courseStartDate = record?.courseDetails?.startDate
-  let traineeStartDate = record?.trainingDetails?.commencementDate
-  if (dates.isInPast(courseStartDate) && !traineeStartDate) {
+exports.traineeDeferred = (record) => {
```
Comment (edwardhorsford) [inline app/lib/utils.js]: There's an existing filter `isDeferred` that returns this.

File: app/routes/existing-record-routes.js [hunk 3]
```diff
@@ -222,6 +222,15 @@ module.exports = router => {
   router.post('/record/:uuid/reinstate/update', (req, res) => {
     const data = req.session.data
     const newRecord = data.record
+
+    let traineeStarted = newRecord?.trainingDetails?.traineeStarted
+
+    // set trainee deferred date as start date if trainee withdrew before starting
```
Comment (edwardhorsford) [inline app/routes/existing-record-routes.js]:
```suggestion
    // Set trainee deferred date as start date if trainee deferred before starting
```

File: app/views/_includes/summary-cards/defer-details.html [hunk 4]
```diff
@@ -14,7 +51,7 @@
           visuallyHiddenText: "deferral date"
         }
       ]
-    } if canAmend
+    } if canAmend and record | traineeStarted
```
Comment (edwardhorsford) [inline app/views/_includes/summary-cards/defer-details.html]: Remind me why you can't edit this?

File: app/views/_includes/summary-cards/trainee-progress.html [hunk 5]
```diff
@@ -126,7 +126,11 @@
     text: "Deferred",
     classes: record.status | getStatusClass + " govuk-!-margin-bottom-2"
   })}} <br>
-  Deferral date: {{record.deferredDate | govukDate}}
+  {% if record.deferredDate %}
+    Deferral date: {{record.deferredDate | govukDate}}
+  {% else %}
+    The trainee deferred before starting their ITT
```
Comment (edwardhorsford) [inline app/views/_includes/summary-cards/trainee-progress.html:132]: Is the deferral date no longer relevant in this case? (just checking)

File: app/views/_includes/summary-cards/trainee-progress.html [hunk 6]
```diff
@@ -241,7 +245,7 @@
         visuallyHiddenText: "trainee’s start date"
       }
     ]
-  } if canAmend and record.courseDetails.startDate | isInPast
+  } if canAmend and record | ittInTheFuture and not record | traineeDeferred
```
Comment (edwardhorsford) [inline app/views/_includes/summary-cards/trainee-progress.html]: Why can you only edit the start date if the ITT is in the future?

**Review by edwardhorsford — CHANGES_REQUESTED** [review]

File: app/lib/utils.js (hunk 1 above)
Comment (ralph-hawkins) [reply app/lib/utils.js:985]: It felt cleaner to use a filter — would you prefer to use `record.courseDetails.startDate | isInFuture`?

File: app/views/_includes/summary-cards/defer-details.html (hunk 4 above)
Comment (ralph-hawkins) [reply app/views/_includes/summary-cards/defer-details.html]: It be pretty complex to resolve on the financial side if the trainee was set to having withdrawn before starting and then it was later changed? 

We could allow it to be edited, but it felt like given the itt had not started or the provider told us that the trainee had not started when the trainee was withdrawn it seemed like we could assume a reasonable amount of intent.

File: app/views/_includes/summary-cards/trainee-progress.html (hunk 5 above)
Comment (ralph-hawkins) [reply app/views/_includes/summary-cards/trainee-progress.html:132]: If the trainee didn't start it the date isn't relevant, only if they did or did not start.

File: app/views/_includes/summary-cards/trainee-progress.html (hunk 6 above)
Comment (ralph-hawkins) [reply app/views/_includes/summary-cards/trainee-progress.html]: I think this should be in the past!

Comment (ralph-hawkins) [reply app/views/_includes/summary-cards/trainee-progress.html]: Fixed it

File: app/lib/utils.js (hunk 1 above)
Comment (edwardhorsford) [reply app/lib/utils.js:985]: It doesn't seem to add too much value to me?

File: app/views/_includes/summary-cards/trainee-progress.html (hunk 5 above)
Comment (edwardhorsford) [reply app/views/_includes/summary-cards/trainee-progress.html:132]: ok!

File: app/lib/utils.js [hunk 7]
```diff
@@ -967,69 +967,62 @@ exports.needsCourseDates = record => {
...
 
-  if (dates.isInFuture(courseStartDate) || !traineeStartDate) {
+  if (dates.isInFuture(ittStartDate)) {
     return true
   } else {
     return false
   }
 }
 
+
```
Comment (edwardhorsford) [inline app/lib/utils.js]: This whole section can be reduced to:
```
return dates.isInFuture(record?.courseDetails?.startDate)
```

File: app/lib/utils.js [hunk 8]
```diff
@@ -967,69 +967,62 @@ exports.needsCourseDates = record => {
...
-exports.showDeferAndWithdraw = (record) => {
-  let courseStartDate = record?.courseDetails?.startDate
-  let status = record?.status
+exports.ittStartedButNoCommencementDate = (record) => {
+  let ittStartDate = record?.courseDetails?.startDate
+  let traineeStartDate = record?.trainingDetails?.commencementDate
 
-  if ((dates.isInPast(courseStartDate)) && (status == "Pending TRN" || status == "TRN received")) {
-    return true 
+  if (dates.isInPast(ittStartDate) && !traineeStartDate) {
```
Comment (edwardhorsford) [inline app/lib/utils.js]:
```suggestion
return (dates.isInPast(ittStartDate) && !traineeStartDate)
```

File: app/lib/utils.js [hunk 9]
```diff
@@ -967,69 +967,62 @@ exports.needsCourseDates = record => {
...
 */
 
-exports.needTraineeStartDate = (record) => {
-  let courseStartDate = record?.courseDetails?.startDate
+exports.traineeStarted = (record) => {
+  let ittStartDate = record?.courseDetails?.startDate
   let traineeStartDate = record?.trainingDetails?.commencementDate
-  if (dates.isInPast(courseStartDate) && !traineeStartDate) {
+
+  if (dates.isInPast(ittStartDate) && traineeStartDate) {
```
Comment (edwardhorsford) [inline app/lib/utils.js]:
```suggestion
 return (dates.isInPast(ittStartDate) && traineeStartDate)
```

**Review by edwardhorsford — APPROVED** [review]
A few of the filters could be simplified - left some suggestions.

**Pull request merged**